api/util/logging.py

```diff
@@ -1,27 +1,73 @@
 import json
 import logging
+import sys
+from datetime import datetime
+from typing import Any
+
+from django.conf import settings
+from gunicorn.config import Config
+from gunicorn.glogging import Logger as GunicornLogger
 
 
 class JsonFormatter(logging.Formatter):
     """Custom formatter for json logs."""
 
-    def format(self, record):
+    def get_json_record(self, record: logging.LogRecord) -> dict[str, Any]:
+        formatted_message = record.getMessage()
+        return {
+            "levelname": record.levelname,
+            "message": formatted_message,
+            "timestamp": self.formatTime(record, self.datefmt),
+            "logger_name": record.name,
+            "process_id": record.process,
+            "thread_name": record.threadName,
+        }
+
+    def format(self, record: logging.LogRecord) -> str:
         """
         %s is replaced with {} because legacy string formatting
         conventions in django-axes module prevent correct
         interpolation of arguments when using this formatter.
         """
         try:
-            log_message = record.msg.replace("%s", "{}")
-            formatted_message = log_message.format(*record.args)
-            log_record = {
-                "levelname": record.levelname,
-                "message": formatted_message,
-                "timestamp": self.formatTime(record, self.datefmt),
-                "logger_name": record.name,
-                "process_id": record.process,
-                "thread_name": record.threadName,
-            }
-            return json.dumps(log_record)
+            return json.dumps(self.get_json_record(record))
         except (ValueError, TypeError) as e:
             return f"Error formatting log record: {str(e)}"
+
+
+class GunicornAccessLogJsonFormatter(JsonFormatter):
+    def get_json_record(self, record: logging.LogRecord) -> dict[str, Any]:
+        response_time = datetime.strptime(record.args["t"], "[%d/%b/%Y:%H:%M:%S %z]")
+        url = record.args["U"]
+        if record.args["q"]:
+            url += f"?{record.args['q']}"
+
+        return {
+            **super().get_json_record(record),
+            "time": response_time.isoformat(),
+            "path": url,
+            "remote_ip": record.args["h"],
+            "method": record.args["m"],
+            "status": str(record.args["s"]),
+            "user_agent": record.args["a"],
+            "referer": record.args["f"],
+            "duration_in_ms": record.args["M"],
+            "pid": record.args["p"],
+        }
+
+
+class GunicornJsonLogger(GunicornLogger):
+    def setup(self, cfg: Config) -> None:
+        super().setup(cfg)
+        if settings.LOG_FORMAT == "json":
```

Comment:
If it's not set to JSON then the logger won't really be a JSON logger, right? It's weird to let it operate as a normal logger with the name of the class `GunicornJsonLogger`

Reply:
Renamed to `GunicornJsonCapableLogger`.